File: frontend/src/services/consolidacao.ts
import { http } from "./http"

export interface ComponentePatrimonio {
  categoria: string
  label: string
  valor_brl: number
  percentual_total: number
  cor: string | null
}

export interface VariacaoPatrimonial {
  valor_atual: number
  valor_referencia: number
  diferenca: number
  variacao_pct: number
}

export interface ConsolidacaoPatrimonial {
  periodo: { ano_id: number; ano: number; mes: number }
  patrimonio_total: number
  componentes: ComponentePatrimonio[]
  total_liquidez: number
  total_renda_fixa: number
  total_renda_variavel: number
  total_cripto: number
  variacao_mes_anterior: VariacaoPatrimonial | null
  variacao_ano_anterior: VariacaoPatrimonial | null
  pct_liquidez: number
  pct_renda_fixa: number
  pct_renda_variavel: number
  pct_cripto: number
  pct_brl: number
  pct_usd: number
  cotacao_usd_brl: number | null
}

export interface ItemPorGeografia {
  geografia: string
  valor_brl: number
  percentual_carteira: number
  qtd_ativos: number
}

export interface ItemPorClasse {
  classe: string
  valor_brl: number
  percentual_carteira: number
  qtd_ativos: number
}

export interface ItemPorAtivo {
  ativo_id: number
  ticker: string
  nome: string
  classe: string
  geografia: string
  valor_brl: number
  percentual_carteira: number
}

export interface ConsolidacaoRV {
  periodo: { ano_id: number; ano: number; mes: number }
  total_brl: number
  total_br_brl: number
  total_eua_brl: number
  total_cripto_brl: number
  por_geografia: ItemPorGeografia[]
  por_classe: ItemPorClasse[]
  por_ativo: ItemPorAtivo[]
}

export const consolidacaoService = {
  patrimonial: (ano_id: number, mes: number) =>
    http.get<ConsolidacaoPatrimonial>("/consolidacao/patrimonial",
      { params: { ano_id, mes } }).then(r => r.data),
  rendaVariavel: (ano_id: number, mes: number) =>
    http.get<ConsolidacaoRV>("/consolidacao/renda-variavel",
      { params: { ano_id, mes } }).then(r => r.data),
}

File: backend/app/schemas/consolidacao.py
from app.schemas.common import ORMBase


# ===================== M21 - Consolidação Renda Variável =====================

class ItemPorClasse(ORMBase):
    classe: str
    valor_brl: float
    percentual_carteira: float
    qtd_ativos: int


class ItemPorGeografia(ORMBase):
    geografia: str
    valor_brl: float
    percentual_carteira: float
    qtd_ativos: int


class ItemPorAtivo(ORMBase):
    ativo_id: int
    ticker: str
    nome: str
    classe: str
    geografia: str
    valor_brl: float
    percentual_carteira: float


class ConsolidacaoRV(ORMBase):
    periodo: dict  # {ano_id, ano, mes}
    total_brl: float
    total_br_brl: float
    total_eua_brl: float
    total_cripto_brl: float
    por_geografia: list[ItemPorGeografia]
    por_classe: list[ItemPorClasse]
    por_ativo: list[ItemPorAtivo]


# ===================== M22 - Consolidação Patrimonial =====================

class ComponentePatrimonio(ORMBase):
    """Cada bloco do patrimônio (contas, RF, previdência, FII...)."""
    categoria: str
    label: str
    valor_brl: float
    percentual_total: float
    cor: str | None = None


class VariacaoPatrimonial(ORMBase):
    valor_atual: float
    valor_referencia: float
    diferenca: float
    variacao_pct: float


class ConsolidacaoPatrimonial(ORMBase):
    periodo: dict
    patrimonio_total: float

    # Componentes do patrimônio
    componentes: list[ComponentePatrimonio]

    # Totais agrupados
    total_liquidez: float          # contas correntes (BRL+USD em BRL)
    total_renda_fixa: float         # produtos RF + previdência + FGTS + fundos
    total_renda_variavel: float    # BR + EUA (em BRL)
    total_cripto: float

    # Comparativos
    variacao_mes_anterior: VariacaoPatrimonial | None = None
    variacao_ano_anterior: VariacaoPatrimonial | None = None

    # Distribuição percentual
    pct_liquidez: float
    pct_renda_fixa: float
    pct_renda_variavel: float
    pct_cripto: float
    pct_brl: float
    pct_usd: float

    # Cotação utilizada
    cotacao_usd_brl: float | None = None